File: Backend/SMA.API/DTOs/ProductDtos.cs
using System.ComponentModel.DataAnnotations;

namespace SMA.API.DTOs
{
    public class CreateProductDto
    {
        [Required]
        public string Sku { get; set; } = string.Empty;

        [Required]
        public string Name { get; set; } = string.Empty;

        public string? Description { get; set; }

        [Range(0.01, 100000)]
        public decimal Price { get; set; }

        [Range(0, 10000)]
        public int InitialStock { get; set; }
    }
    public class ProductResponseDto
    {
        public Guid Id { get; set; }
        public string Sku { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public string? Description { get; set; }
        public decimal Price { get; set; }
        public int QuantityAvailable { get; set; } // Flattened from the Inventory table
        public double AverageRating { get; set; }
        public int RatingCount { get; set; }
    }

    public class ProductRatingSummaryDto
    {
        public double AverageRating { get; set; }
        public int RatingCount { get; set; }
        public int? CurrentUserRating { get; set; }
        public bool CanRate { get; set; }
    }

    public class CreateProductRatingDto
    {
        [Range(1, 5)]
        public int Rating { get; set; }

        [MaxLength(2000)]
        public string? Feedback { get; set; }
    }

    public class ProductRatingResponseDto
    {
        public Guid Id { get; set; }
        public Guid ProductId { get; set; }
        public string ProductName { get; set; } = string.Empty;
        public string CustomerName { get; set; } = string.Empty;
        public string CustomerEmail { get; set; } = string.Empty;
        public int Rating { get; set; }
        public string? Feedback { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
    }

    public class PagedProductRatingResponseDto
    {
        public List<ProductRatingResponseDto> Items { get; set; } = [];
        public int Page { get; set; }
        public int PageSize { get; set; }
        public int TotalCount { get; set; }
        public int TotalPages { get; set; }
    }

    public class AdminProductResponseDto
    {
        public Guid Id { get; set; }
        public string Sku { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public string? Description { get; set; }
        public decimal Price { get; set; }
        public bool IsActive { get; set; }
        public int QuantityAvailable { get; set; }
        public int QuantityReserved { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
    }

    public class UpdateProductDto
    {
        [MaxLength(50)]
        public string? Sku { get; set; }

        [MaxLength(255)]
        public string? Name { get; set; }

        public string? Description { get; set; }

        [Range(0.01, 100000)]
        public decimal? Price { get; set; }

        public bool? IsActive { get; set; }
    }

    public class UpdateStockDto
    {
        [Range(0, 100000)]
        public int QuantityAvailable { get; set; }
    }
}

File: Frontend/sma-ui/DTOs/ProductDTOs.ts
export interface ProductResponseDto {
  id: string;
  sku: string;
  name: string;
  description: string | null;
  price: number;
  quantityAvailable: number;
  averageRating: number;
  ratingCount: number;
}

export interface ProductRatingSummaryDto {
  averageRating: number;
  ratingCount: number;
  currentUserRating: number | null;
  canRate: boolean;
}

export interface CreateProductRatingDto {
  rating: number;
  feedback?: string;
}

export interface ProductRatingResponseDto {
  id: string;
  productId: string;
  productName: string;
  customerName: string;
  customerEmail: string;
  rating: number;
  feedback: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface PagedProductRatingResponseDto {
  items: ProductRatingResponseDto[];
  page: number;
  pageSize: number;
  totalCount: number;
  totalPages: number;
}

export interface AdminProductResponseDto {
  id: string;
  sku: string;
  name: string;
  description: string | null;
  price: number;
  isActive: boolean;
  quantityAvailable: number;
  quantityReserved: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateProductDto {
  sku: string;
  name: string;
  description?: string;
  price: number;
  initialStock: number;
}

export interface UpdateProductDto {
  sku?: string;
  name?: string;
  description?: string;
  price?: number;
  isActive?: boolean;
}

export interface UpdateStockDto {
  quantityAvailable: number;
}
